File: test/fixtures/vulnerable-app.js
// Intentionally vulnerable fixture. Test corpus only, never runnable code.
// Use this to test sentinel-mcp: scan_file({ filePath: "/path/to/this/file" })

const express = require("express");
const jwt = require("jsonwebtoken");
const bcrypt = require("bcryptjs");
const { exec } = require("child_process");
const fs = require("fs");

const app = express();

// SECRET: Hardcoded credentials
const API_KEY = "sk_live_notarealkeyusedasfix";
const dbPassword = "super_secret_database_password_123!";
const JWT_SECRET = "my-jwt-secret-that-should-be-in-env";
const awsKey = "AKIAIOSFODNN7EXAMPLE";
const privateKey = "-----BEGIN RSA PRIVATE KEY-----\nMIIEpAIBAAKCAQ...";

// SQL INJECTION
app.get("/users", async (req, res) => {
  const userId = req.query.id;
  const result = await db.query(`SELECT * FROM users WHERE id = ${userId}`);
  res.json(result);
});

app.post("/search", async (req, res) => {
  const name = req.body.name;
  const result = await db.query("SELECT * FROM products WHERE name = '" + name + "'");
  res.json(result);
});

// COMMAND INJECTION
app.get("/ping", (req, res) => {
  const host = req.query.host;
  exec(`ping -c 1 ${host}`, (err, stdout) => {
    res.send(stdout);
  });
});

// XSS
app.get("/greet", (req, res) => {
  const name = req.query.name;
  res.send(`<h1>Hello ${name}</h1>`);
});

app.get("/profile", (req, res) => {
  document.getElementById("output").innerHTML = userData;
});

// EVAL
app.post("/calculate", (req, res) => {
  const expression = req.body.expr;
  const result = eval(expression);
  res.json({ result });
});

// PATH TRAVERSAL
app.get("/file", (req, res) => {
  const filename = req.query.name;
  const content = fs.readFileSync("/uploads/" + filename, "utf-8");
  res.send(content);
});

// WEAK CRYPTO
const crypto = require("crypto");
function hashPassword(password) {
  return crypto.createHash("md5").update(password).digest("hex");
}

function generateToken() {
  return Math.random().toString(36).substring(2);
}

// JWT ISSUES
app.post("/login", async (req, res) => {
  const user = await findUser(req.body.email);
  const passwordMatch = await bcrypt.hash(req.body.password, 4); // Low rounds

  // Hardcoded JWT secret
  const token = jwt.sign({ userId: user.id, role: user.role }, "hardcoded-secret-key");
  res.json({ token });
});

app.get("/profile", (req, res) => {
  // Decoding without verification!
  const decoded = jwt.decode(req.headers.authorization);
  res.json(decoded);
});

// PROTOTYPE POLLUTION
function merge(target, source) {
  for (const key in source) {
    if (typeof source[key] === "object") {
      target[key] = merge(target[key] || {}, source[key]);
    } else {
      target[key] = source[key];
    }
  }
  return target;
}

app.post("/settings", (req, res) => {
  const settings = {};
  merge(settings, req.body); // Prototype pollution!
  res.json(settings);
});

// OPEN REDIRECT
app.get("/redirect", (req, res) => {
  const url = req.query.url;
  res.redirect(req.query.url);
});

// NOSQL INJECTION
app.post("/api/login", async (req, res) => {
  const user = await User.findOne(req.body);
  res.json(user);
});

// CORS MISCONFIGURATION
app.use((req, res, next) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  next();
});

// INFO DISCLOSURE
app.use((err, req, res, next) => {
  res.send(err.stack);
});

app.listen(3000);
console.log("Server running on http://localhost:3000");

File: test/fixtures/vulnerable-app.py
# Intentionally vulnerable fixture. Test corpus only, never runnable code.
# Use this to test: scan_file({ filePath: "/path/to/this/file" })

import os
import pickle
import yaml
import hashlib
import random
import subprocess
import sqlite3
from flask import Flask, request, render_template_string, send_file, redirect

app = Flask(__name__)

# SECRETS
SECRET_KEY = "super-secret-flask-key-that-should-be-in-env"
app.secret_key = "another-hardcoded-secret-key"
API_KEY = "sk_live_notarealkeyusedasfix"
DATABASE_URL = "postgresql://admin:password123@db.example.com:5432/prod"
AWS_KEY = "AKIAIOSFODNN7EXAMPLE"

# SQL INJECTION
@app.route("/users")
def get_user():
    user_id = request.args.get("id")
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM users WHERE id = {user_id}")
    return str(cursor.fetchall())

@app.route("/search")
def search():
    query = request.args.get("q")
    cursor.execute("SELECT * FROM products WHERE name = '%s'" % query)
    return str(cursor.fetchall())

# COMMAND INJECTION
@app.route("/ping")
def ping():
    host = request.args.get("host")
    os.system(f"ping -c 1 {host}")
    return "Done"

@app.route("/convert")
def convert():
    filename = request.args.get("file")
    subprocess.call(f"convert {filename} output.pdf", shell=True)
    return "Converted"

@app.route("/legacy")
def legacy():
    cmd = request.args.get("cmd")
    result = os.popen(cmd).read()
    return result

# INSECURE DESERIALIZATION
@app.route("/load", methods=["POST"])
def load_data():
    data = request.get_data()
    obj = pickle.loads(data)
    return str(obj)

@app.route("/config")
def load_config():
    config_data = request.args.get("config")
    config = yaml.load(config_data)
    return str(config)

# XSS / TEMPLATE INJECTION
@app.route("/greet")
def greet():
    name = request.args.get("name")
    return render_template_string(f"<h1>Hello {name}</h1>")

# PATH TRAVERSAL
@app.route("/download")
def download():
    filename = request.args.get("file")
    return send_file(f"/uploads/{filename}")

@app.route("/read")
def read_file():
    path = request.args.get("path")
    content = open(f"/data/{path}").read()
    return content

# WEAK CRYPTO
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()

def generate_token():
    return str(random.randint(100000, 999999))

def hash_data(data):
    return hashlib.sha1(data.encode()).hexdigest()

# AUTH ISSUES
import jwt

def get_user_from_token(token):
    # No verification!
    data = jwt.decode(token, options={"verify_signature": False})
    return data

def check_admin(request):
    assert request.user.is_admin  # assert stripped with -O flag!

# DJANGO-STYLE ISSUES
DEBUG = True
ALLOWED_HOSTS = ['*']

# FLASK DEBUG IN PRODUCTION
if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0")
